File: packages/virage-core/src/quality/metrics/dense-embedding.ts
/**
 * Component 4 — Dense Embedding metrics (5 metrics)
 *
 * Self-Recall@K:      chunk-as-query retrieval hit rate (must-pass >0.80).
 * Intrinsic Dimension: PCA components to explain 95% variance (target 70–90% of dims).
 * Uniformity:         kd-tree kth-nearest distance evenness (target 0.7–0.85).
 * Isotropy:           min/max eigenvalue ratio of covariance matrix.
 * Outlier Fraction:   fraction with no close neighbours (must-pass <0.05).
 */

import type { MetricResult } from "../interfaces.js";
import {
  normalizeMonotonicUp01,
  normalizeMonotonicDown,
  normalizeIntrinsicDimension,
  normalizeUniformity,
} from "../scoring.js";

export interface DenseEmbeddingMetricsInput {
  chunks: Array<{ id: string; denseText: string; anchorText?: string }>;
  searchFn: (query: string, topK: number) => Promise<Array<{ id: string }>>;
  embedFn: (text: string) => Promise<number[]>;
  topK: number;
}

// ─── Self-Recall@K ─────────────────────────────────────────────────────────────

async function computeSelfRecall(
  chunks: DenseEmbeddingMetricsInput["chunks"],
  searchFn: DenseEmbeddingMetricsInput["searchFn"],
  topK: number,
): Promise<number> {
  if (chunks.length === 0) return 0;
  let hits = 0;
  for (const chunk of chunks) {
    const query = chunk.anchorText ?? chunk.denseText.slice(0, 80);
    const results = await searchFn(query, topK);
    if (results.some((r) => r.id === chunk.id)) hits++;
  }
  return hits / chunks.length;
}

// ─── PCA helpers (no dependency on external libs) ─────────────────────────────

function mean(values: number[]): number {
  return values.reduce((s, v) => s + v, 0) / values.length;
}

/**
 * Power iteration to find the largest eigenvector/eigenvalue of a symmetric matrix.
 * Returns [eigenvalue, eigenvector].
 */
function powerIteration(M: number[][], maxIter = 100): [number, number[]] {
  const n = M.length;
  let v = Array.from({ length: n }, () => Math.random() - 0.5);
  let lambda = 0;
  for (let iter = 0; iter < maxIter; iter++) {
    const Mv = M.map((row) => row.reduce((s, val, j) => s + val * v[j], 0));
    lambda = Math.sqrt(Mv.reduce((s, x) => s + x * x, 0));
    if (lambda < 1e-10) break;
    v = Mv.map((x) => x / lambda);
  }
  return [lambda, v];
}

/**
 * Lightweight PCA: uses repeated power iteration with deflation.
 * Returns explained variance fractions for each component.
 * Suitable for sampleSize <= 500, dims <= 1024.
 */
function pca(matrix: number[][], numComponents: number): number[] {
  const n = matrix.length;
  const d = matrix[0].length;
  const k = Math.min(numComponents, d, n - 1);

  const colMeans = Array.from({ length: d }, (_, j) =>
    mean(matrix.map((row) => row[j])),
  );
  const centered = matrix.map((row) => row.map((v, j) => v - colMeans[j]));

  // Covariance matrix (d × d)
  const cov: number[][] = Array.from({ length: d }, () => Array(d).fill(0));
  for (const row of centered) {
    for (let i = 0; i < d; i++) {
      for (let j = 0; j < d; j++) {
        cov[i][j] += (row[i] * row[j]) / n;
      }
    }
  }

  const eigenvalues: number[] = [];
  const residual = cov.map((r) => [...r]);

  for (let c = 0; c < k; c++) {
    const [lambda, v] = powerIteration(residual);
    eigenvalues.push(lambda);
    // Deflate: residual = residual - lambda * v * v^T
    for (let i = 0; i < d; i++) {
      for (let j = 0; j < d; j++) {
        residual[i][j] -= lambda * v[i] * v[j];
      }
    }
  }

  const totalVar = eigenvalues.reduce((s, e) => s + e, 0);
  if (totalVar === 0) return eigenvalues.map(() => 0);
  return eigenvalues.map((e) => e / totalVar);
}

function computeIntrinsicDimension(vectors: number[][]): {
  fraction: number;
  components95: number;
  totalDims: number;
} {
  if (vectors.length < 4) {
    return { fraction: 0, components95: 0, totalDims: vectors[0]?.length ?? 0 };
  }
  const totalDims = vectors[0].length;
  const maxComponents = Math.min(totalDims, vectors.length - 1, 64);
  const variances = pca(vectors, maxComponents);
  let cumVar = 0;
  let components95 = maxComponents;
  for (let i = 0; i < variances.length; i++) {
    cumVar += variances[i];
    if (cumVar >= 0.95) {
      components95 = i + 1;
      break;
    }
  }
  return {
    fraction: components95 / totalDims,
    components95,
    totalDims,
  };
}

// ─── Uniformity (kth-nearest-distance evenness) ───────────────────────────────

function l2(a: number[], b: number[]): number {
  return Math.sqrt(a.reduce((s, v, i) => s + (v - b[i]) ** 2, 0));
}

function computeUniformity(vectors: number[][], k = 5): number {
  if (vectors.length < k + 1) return 0;
  const kthDistances = vectors.map((v) => {
    const dists = vectors
      .filter((u) => u !== v)
      .map((u) => l2(v, u))
      .sort((a, b) => a - b);
    return dists[k - 1] ?? 0;
  });
  const avg = mean(kthDistances);
  const stdDev = Math.sqrt(mean(kthDistances.map((d) => (d - avg) ** 2)));
  if (avg === 0) return 0;
  return Math.max(0, Math.min(1, 1 - stdDev / avg));
}

// ─── Isotropy (min/max eigenvalue ratio) ─────────────────────────────────────

function computeIsotropy(vectors: number[][]): number {
  if (vectors.length < 4) return 0;
  const d = vectors[0].length;
  const n = vectors.length;
  const colMeans = Array.from({ length: d }, (_, j) =>
    mean(vectors.map((row) => row[j])),
  );
  const centered = vectors.map((row) => row.map((v, j) => v - colMeans[j]));
  const cov: number[][] = Array.from({ length: d }, () => Array(d).fill(0));
  for (const row of centered) {
    for (let i = 0; i < d; i++) {
      for (let j = 0; j < d; j++) {
        cov[i][j] += (row[i] * row[j]) / n;
      }
    }
  }

  const maxComponents = Math.min(d, n - 1, 32);
  const variances: number[] = [];
  const residual = cov.map((r) => [...r]);
  for (let c = 0; c < maxComponents; c++) {
    const [lambda] = powerIteration(residual);
    variances.push(lambda);
    const [, v] = powerIteration(residual);
    for (let i = 0; i < d; i++) {
      for (let j = 0; j < d; j++) {
        residual[i][j] -= lambda * v[i] * v[j];
      }
    }
  }

  if (variances.length < 2) return 0;
  const maxEig = Math.max(...variances);
  const minEig = Math.min(...variances.filter((v) => v > 1e-10));
  return maxEig === 0 ? 0 : Math.min(1, minEig / maxEig);
}

// ─── Outlier Fraction ─────────────────────────────────────────────────────────

function median(sorted: number[]): number {
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0
    ? sorted[mid]
    : (sorted[mid - 1] + sorted[mid]) / 2;
}

function computeOutlierFraction(vectors: number[][]): number {
  if (vectors.length < 4) return 0;
  const nearestDists = vectors.map((v) => {
    const dists = vectors.filter((u) => u !== v).map((u) => l2(v, u));
    return Math.min(...dists);
  });
  const sorted = [...nearestDists].sort((a, b) => a - b);
  const med = median(sorted);
  const mad = median(
    sorted.map((d) => Math.abs(d - med)).sort((a, b) => a - b),
  );
  const threshold = med + 3 * mad;
  const outliers = nearestDists.filter((d) => d > threshold);
  return outliers.length / vectors.length;
}

// ─── Main ─────────────────────────────────────────────────────────────────────

export async function computeDenseEmbeddingMetrics(
  input: DenseEmbeddingMetricsInput,
  sampleSize = 200,
  selfRecallThreshold = 0.8,
  outlierThreshold = 0.05,
  weightOverrides: Partial<Record<string, number>> = {},
): Promise<MetricResult[]> {
  const { chunks, searchFn, embedFn, topK } = input;
  const sample = chunks.slice(0, sampleSize);

  // Self-Recall (must pass >0.80)
  const selfRecall = await computeSelfRecall(sample, searchFn, topK);

  // Embed the sample for geometry metrics
  const vectors = await Promise.all(sample.map((c) => embedFn(c.denseText)));

  const { fraction: idFraction } = computeIntrinsicDimension(vectors);
  const uniformity = computeUniformity(vectors);
  const isotropy = computeIsotropy(vectors);
  const outlierFraction = computeOutlierFraction(vectors);

  return [
    {
      name: "SelfRecall@K",
      rawValue: selfRecall,
      normalizedValue: normalizeMonotonicUp01(selfRecall),
      weight: weightOverrides["selfRecall"] ?? 3.0,
      skipped: false,
      mustPass: true,
      mustPassThreshold: selfRecallThreshold,
      mustPassPassed: selfRecall > selfRecallThreshold,
    },
    {
      name: "IntrinsicDimension",
      rawValue: idFraction,
      normalizedValue: normalizeIntrinsicDimension(idFraction),
      weight: weightOverrides["intrinsicDimension"] ?? 1.0,
      skipped: vectors.length < 4,
      skipReason:
        vectors.length < 4 ? "Insufficient sample for PCA" : undefined,
    },
    {
      name: "Uniformity",
      rawValue: uniformity,
      normalizedValue: normalizeUniformity(uniformity),
      weight: weightOverrides["uniformity"] ?? 1.0,
      skipped: vectors.length < 4,
      skipReason: vectors.length < 4 ? "Insufficient sample" : undefined,
    },
    {
      name: "Isotropy",
      rawValue: isotropy,
      normalizedValue: normalizeMonotonicUp01(isotropy),
      weight: weightOverrides["isotropy"] ?? 0.5,
      skipped: vectors.length < 4,
      skipReason: vectors.length < 4 ? "Insufficient sample" : undefined,
    },
    {
      name: "OutlierFraction",
      rawValue: outlierFraction,
      normalizedValue: normalizeMonotonicDown(outlierFraction),
      weight: weightOverrides["outlierFraction"] ?? 1.0,
      skipped: vectors.length < 4,
      skipReason: vectors.length < 4 ? "Insufficient sample" : undefined,
      mustPass: true,
      mustPassThreshold: outlierThreshold,
      mustPassPassed: outlierFraction < outlierThreshold,
    },
  ];
}

File: crates/virage-engine/src/quality/metrics/dense_embedding.rs
//! Component 4 — Dense Embedding metrics (5 metrics).
//!
//! Self-Recall@K:      chunk-as-query retrieval hit rate (must-pass). Computed by the
//!                     runner (needs async vector-store search) and passed in as `self_recall`.
//! Intrinsic Dimension: PCA components to explain 95% variance (target 70-90% of dims).
//! Uniformity:         kd-tree kth-nearest distance evenness (target 0.7-0.85).
//! Isotropy:           min/max eigenvalue ratio of covariance matrix.
//! Outlier Fraction:   fraction with no close neighbours (must-pass).
//!
//! Ported from `dist/quality/metrics/dense-embedding.js` — IR-038.

use crate::quality::scoring::{
    normalize_intrinsic_dimension, normalize_monotonic_down, normalize_monotonic_up01,
    normalize_uniformity,
};
use crate::quality::MetricResult;

use super::{l2, mean, median, Lcg};

/// Power iteration to find the largest eigenvalue/eigenvector of a symmetric matrix.
fn power_iteration(m: &[Vec<f64>], rng: &mut Lcg, max_iter: usize) -> (f64, Vec<f64>) {
    let n = m.len();
    let mut v: Vec<f64> = (0..n).map(|_| rng.next_unit()).collect();
    let mut lambda = 0.0;
    for _ in 0..max_iter {
        let mv: Vec<f64> = m
            .iter()
            .map(|row| row.iter().zip(v.iter()).map(|(a, b)| a * b).sum::<f64>())
            .collect();
        lambda = mv.iter().map(|x| x * x).sum::<f64>().sqrt();
        if lambda < 1e-10 {
            break;
        }
        v = mv.iter().map(|x| x / lambda).collect();
    }
    (lambda, v)
}

fn covariance_matrix(matrix: &[Vec<f32>]) -> (usize, Vec<Vec<f64>>) {
    let n = matrix.len();
    let d = matrix[0].len();
    let col_means: Vec<f64> = (0..d)
        .map(|j| mean(&matrix.iter().map(|row| row[j] as f64).collect::<Vec<_>>()))
        .collect();
    let centered: Vec<Vec<f64>> = matrix
        .iter()
        .map(|row| {
            row.iter()
                .enumerate()
                .map(|(j, v)| *v as f64 - col_means[j])
                .collect()
        })
        .collect();
    let mut cov = vec![vec![0.0; d]; d];
    for row in &centered {
        for i in 0..d {
            for j in 0..d {
                cov[i][j] += (row[i] * row[j]) / n as f64;
            }
        }
    }
    (d, cov)
}

/// Lightweight PCA via repeated power iteration with deflation.
/// Returns explained variance fractions for each component.
fn pca(matrix: &[Vec<f32>], num_components: usize, rng: &mut Lcg) -> Vec<f64> {
    let n = matrix.len();
    let (d, cov) = covariance_matrix(matrix);
    let k = num_components.min(d).min(n.saturating_sub(1));
    let mut residual = cov;
    let mut eigenvalues = Vec::with_capacity(k);
    for _ in 0..k {
        let (lambda, v) = power_iteration(&residual, rng, 100);
        eigenvalues.push(lambda);
        for i in 0..d {
            for j in 0..d {
                residual[i][j] -= lambda * v[i] * v[j];
            }
        }
    }
    let total_var: f64 = eigenvalues.iter().sum();
    if total_var == 0.0 {
        eigenvalues.iter().map(|_| 0.0).collect()
    } else {
        eigenvalues.iter().map(|e| e / total_var).collect()
    }
}

fn compute_intrinsic_dimension(vectors: &[Vec<f32>], rng: &mut Lcg) -> f64 {
    if vectors.len() < 4 {
        return 0.0;
    }
    let total_dims = vectors[0].len();
    let max_components = total_dims.min(vectors.len() - 1).min(64);
    let variances = pca(vectors, max_components, rng);
    let mut cum_var = 0.0;
    let mut components95 = max_components;
    for (i, v) in variances.iter().enumerate() {
        cum_var += v;
        if cum_var >= 0.95 {
            components95 = i + 1;
            break;
        }
    }
    components95 as f64 / total_dims as f64
}

fn compute_uniformity(vectors: &[Vec<f32>], k: usize) -> f64 {
    if vectors.len() < k + 1 {
        return 0.0;
    }
    let kth_distances: Vec<f64> = vectors
        .iter()
        .map(|v| {
            let mut dists: Vec<f64> = vectors
                .iter()
                .filter(|u| !std::ptr::eq(u.as_slice(), v.as_slice()))
                .map(|u| l2(v, u))
                .collect();
            dists.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
            dists.get(k - 1).copied().unwrap_or(0.0)
        })
        .collect();
    let avg = mean(&kth_distances);
    let std_dev = mean(
        &kth_distances
            .iter()
            .map(|d| (d - avg).powi(2))
            .collect::<Vec<_>>(),
    )
    .sqrt();
    if avg == 0.0 {
        0.0
    } else {
        (1.0 - std_dev / avg).clamp(0.0, 1.0)
    }
}

fn compute_isotropy(vectors: &[Vec<f32>], rng: &mut Lcg) -> f64 {
    if vectors.len() < 4 {
        return 0.0;
    }
    let n = vectors.len();
    let (d, cov) = covariance_matrix(vectors);
    let max_components = d.min(n - 1).min(32);
    let mut residual = cov;
    let mut variances = Vec::with_capacity(max_components);
    for _ in 0..max_components {
        let (lambda, v) = power_iteration(&residual, rng, 100);
        variances.push(lambda);
        for i in 0..d {
            for j in 0..d {
                residual[i][j] -= lambda * v[i] * v[j];
            }
        }
    }
    if variances.len() < 2 {
        return 0.0;
    }
    let max_eig = variances.iter().cloned().fold(f64::MIN, f64::max);
    let min_eig = variances
        .iter()
        .cloned()
        .filter(|v| *v > 1e-10)
        .fold(f64::MAX, f64::min);
    if max_eig == 0.0 {
        0.0
    } else {
        (min_eig / max_eig).min(1.0)
    }
}

fn compute_outlier_fraction(vectors: &[Vec<f32>]) -> f64 {
    if vectors.len() < 4 {
        return 0.0;
    }
    let nearest_dists: Vec<f64> = vectors
        .iter()
        .map(|v| {
            vectors
                .iter()
                .filter(|u| !std::ptr::eq(u.as_slice(), v.as_slice()))
                .map(|u| l2(v, u))
                .fold(f64::MAX, f64::min)
        })
        .collect();
    let mut sorted = nearest_dists.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let med = median(&sorted);
    let mut abs_devs: Vec<f64> = sorted.iter().map(|d| (d - med).abs()).collect();
    abs_devs.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let mad = median(&abs_devs);
    let threshold = med + 3.0 * mad;
    let outliers = nearest_dists.iter().filter(|d| **d > threshold).count();
    outliers as f64 / vectors.len() as f64
}

#[allow(clippy::too_many_arguments)]
pub fn compute_dense_embedding_metrics(
    self_recall: f64,
    vectors: &[Vec<f32>],
    self_recall_threshold: f64,
    outlier_threshold: f64,
) -> Vec<MetricResult> {
    let mut rng = Lcg(0xC0FFEE_u64);
    let insufficient = vectors.len() < 4;
    let id_fraction = compute_intrinsic_dimension(vectors, &mut rng);
    let uniformity = compute_uniformity(vectors, 5);
    let isotropy = compute_isotropy(vectors, &mut rng);
    let outlier_fraction = compute_outlier_fraction(vectors);

    let mut results = Vec::with_capacity(5);
    results.push(
        MetricResult::new(
            "SelfRecall@K",
            self_recall,
            normalize_monotonic_up01(self_recall),
            3.0,
        )
        .with_must_pass(self_recall_threshold, self_recall > self_recall_threshold),
    );
    if insufficient {
        results.push(MetricResult::skipped(
            "IntrinsicDimension",
            1.0,
            "Insufficient sample for PCA",
        ));
        results.push(MetricResult::skipped(
            "Uniformity",
            1.0,
            "Insufficient sample",
        ));
        results.push(MetricResult::skipped(
            "Isotropy",
            0.5,
            "Insufficient sample",
        ));
        let mut m = MetricResult::skipped("OutlierFraction", 1.0, "Insufficient sample");
        m.must_pass = true;
        m.must_pass_threshold = Some(outlier_threshold);
        results.push(m);
    } else {
        results.push(MetricResult::new(
            "IntrinsicDimension",
            id_fraction,
            normalize_intrinsic_dimension(id_fraction),
            1.0,
        ));
        results.push(MetricResult::new(
            "Uniformity",
            uniformity,
            normalize_uniformity(uniformity),
            1.0,
        ));
        results.push(MetricResult::new(
            "Isotropy",
            isotropy,
            normalize_monotonic_up01(isotropy),
            0.5,
        ));
        results.push(
            MetricResult::new(
                "OutlierFraction",
                outlier_fraction,
                normalize_monotonic_down(outlier_fraction),
                1.0,
            )
            .with_must_pass(outlier_threshold, outlier_fraction < outlier_threshold),
        );
    }
    results
}
